src/ExactDiagonalization/exact_diagonalization_problem.jl

```diff
@@ -111,17 +117,68 @@ See also [`solve(::ExactDiagonalizationProblem)`](@ref),
     Using the `LOBPCGSolver()` algorithm requires the IterativeSolvers.jl package. The package
     can be loaded with `using IterativeSolvers`.
 """
-struct ExactDiagonalizationProblem{H<:AbstractHamiltonian, V}
+struct ExactDiagonalizationProblem{H<:AbstractHamiltonian,V,ALG<:AbstractAlgorithm,AV}
     hamiltonian::H
     initial_vector::V
+    algorithm::ALG
+    linear_dimension::Number
+    addr_or_vec::AV # starting address or iterable of addresses
     kwargs::NamedTuple
+end
 
-    function ExactDiagonalizationProblem(
-        hamiltonian::H, initial_vector::V=nothing; kwargs...
-    ) where {H<:AbstractHamiltonian,V}
-        return new{H,V}(hamiltonian, initial_vector, NamedTuple(kwargs))
+function ExactDiagonalizationProblem(
+    hamiltonian::H, initial_vector::V=nothing;
+    linear_dimension=nothing, algorithm::ALG=LinearAlgebraSolver(),
+    info=false, warn=true, kwargs...
+) where {H<:AbstractHamiltonian,V,ALG<:AbstractAlgorithm}
+    # Set up the starting address or vector of addresses
+    addr_or_vec = _set_up_starting_address(initial_vector, hamiltonian)
+    if linear_dimension === nothing
+        linear_dimension = dimension(
+            hamiltonian,
+            addr_or_vec isa AbstractFockAddress ? addr_or_vec : first(addr_or_vec)
+        )
+    end
+    dense, matrix_free, sparse = estimate_memory_requirement(
+        hamiltonian, linear_dimension
+    )
+    free_memory = Sys.free_memory() # available memory in bytes
+    total_memory = Sys.total_memory() # total memory in bytes
```

Comment:
We should be careful with stuff like this: free memory may not actually reflect the amount of available memory as either the OS kernel or Julia's GC might free up some memory when needed.

Reply:
Maybe just ignore the `free_memory`. The `total_memory` should be more significant. There is also a `Sys.total_physical_memory()`.